src/allocators/allocator-guarantees.md

```diff
@@ -0,0 +1,58 @@
+## Recommendation for allocator implementations
+
+### Allocating memory
+
+In addition to implementing the conventational invariants ensuring the
+mutually exclusive allocation of memory, CHERI-aware implementations of
+`malloc()` and `calloc()` must return a capability that has the following
+properties:
+
+ * Is valid (i.e., with its tag bit set)
+ * Is unsealed
+ * Has bounds that permit access to the full requested range of the allocation
+ * Has bounds that do not permit access to any other current allocation,
+   implementing non-aliasing spatial safety
+ * Has permissions that allow data load, data store, capability load, and
+   capability store
+ * Be sufficiently aligned to allow capability loads and stores at relative
+   offset 0 from the returned pointer
+
+The allocator should:
+
+ * Zero the allocation before returning a pointer to it
+
+The allocator may:
+
+ * Provide precise bounds -- i.e., in which the lower bound is the lowest
+   address of the returned allocation, and the upper bound is the highest
+   address of the returned allocation plus one
+
+### Freeing memory
+
+CHERI-aware allocators must ignore, or trap on, calls to `free()` if the
+passed capability:
+
+ * Is invalid (i.e, with its tag bit unset); or
+ * Is sealed; or
+ * Has bounds that have been changed from those returned by the allocator
+   when the allocation was first made; or
+ * Has permissions that disallow any of data load, data store, capability
+   load, or capability store
+
+The allocator may implement fail-stop semantics if the call fails for one or
+more of the above reasons.
+
```

Comment:
It is probably safe to leave up to the allocator whether a nonzero offset between address and base is acceptable.

The CHERIoT allocator overloads `free()` (well, `heap_free()`) to permit dropping claims on an object, and in that case it accepts interior, subset, &| attenuated pointers.  That is, claims may be (taken and) dropped on subobjects (or, really, arbitrary slices of objects) but deallocation requires the original pointer (a la C).

Reply:
Regarding the former: Are there use cases for non-zero offsets in heap allocators we are aware of beyond, perhaps, "deterministically trap on overflow rather than on underflow when imprecise bounds require padding"?